sys/net/routing/ng_rpl/ng_rpl_control_messages.c

```diff
@@ -0,0 +1,579 @@
+    if (!ng_netapi_dispatch_send(NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL,hdr)) {
+        DEBUG("RPL: cannot send packet: no subscribers found.\n");
+        ng_pktbuf_release(hdr);
+    }
+
+}
+
+void ng_rpl_send_DIO(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *destination)
+{
+    if (dodag == NULL) {
+        DEBUG("RPL: Error - trying to send DIO without being part of a dodag.\n");
+        return;
+    }
+
+    ng_pktsnip_t *pkt;
+    ng_icmpv6_hdr_t *icmp;
+    ng_rpl_dio_t *dio;
+    ng_rpl_opt_dodag_conf_t *dodag_conf;
+    uint8_t *pos;
+    int size = sizeof(ng_icmpv6_hdr_t) + sizeof(ng_rpl_dio_t);
+
+    if ((dodag->dodag_conf_counter % 3) == 0) {
+        size += sizeof(ng_rpl_opt_dodag_conf_t);
+    }
+
+    if ((pkt = ng_icmpv6_build(NULL, NG_ICMPV6_RPL_CTRL, NG_RPL_ICMPV6_CODE_DIO, size)) == NULL) {
+        DEBUG("RPL: Send DIO - no space left in packet buffer\n");
+        return;
+    }
+
+    icmp = (ng_icmpv6_hdr_t *)pkt->data;
+    dio = (ng_rpl_dio_t *)(icmp + 1);
+    pos = (uint8_t *) dio;
+    dio->instance_id = dodag->instance->id;
+    dio->version_number = dodag->version;
+    dio->rank = byteorder_htons(dodag->my_rank);
+    dio->g_mop_prf = (dodag->grounded << NG_RPL_GROUNDED_SHIFT) |
+        (dodag->instance->mop << NG_RPL_MOP_SHIFT) | dodag->prf;
+    dio->dtsn = dodag->dtsn;
+    dio->flags = 0;
+    dio->reserved = 0;
+    dio->dodag_id = dodag->dodag_id;
+
+    pos += sizeof(*dio);
+
+    if ((dodag->dodag_conf_counter % 3) == 0) {
+        dodag_conf = (ng_rpl_opt_dodag_conf_t *) pos;
+        dodag_conf->type = NG_RPL_OPT_DODAG_CONF;
+        dodag_conf->length = NG_RPL_OPT_DODAG_CONF_LEN;
+        dodag_conf->flags_a_pcs = 0;
+        dodag_conf->dio_int_doubl = dodag->dio_interval_doubl;
+        dodag_conf->dio_int_min = dodag->dio_min;
+        dodag_conf->dio_redun = dodag->dio_redun;
+        dodag_conf->max_rank_inc = byteorder_htons(dodag->instance->max_rank_inc);
+        dodag_conf->min_hop_rank_inc = byteorder_htons(dodag->instance->min_hop_rank_inc);
+        dodag_conf->ocp = byteorder_htons(dodag->instance->of->ocp);
+        dodag_conf->reserved = 0;
+        dodag_conf->default_lifetime = dodag->default_lifetime;
+        dodag_conf->lifetime_unit = byteorder_htons(dodag->lifetime_unit);
+        pos += sizeof(*dodag_conf);
+    }
+
+    dodag->dodag_conf_counter++;
+    _ng_rpl_send(pkt, NULL, destination);
+}
+
+void ng_rpl_send_DIS(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *destination)
+{
+    (void) dodag;
+    ng_pktsnip_t *pkt;
+    ng_icmpv6_hdr_t *icmp;
+    ng_rpl_dis_t *dis;
+    int size = sizeof(ng_icmpv6_hdr_t) + sizeof(ng_rpl_dis_t) + 4;
+
+    if ((pkt = ng_icmpv6_build(NULL, NG_ICMPV6_RPL_CTRL, NG_RPL_ICMPV6_CODE_DIS, size)) == NULL) {
+        DEBUG("RPL: Send DIS - no space left in packet buffer\n");
+        return;
+    }
+
+    icmp = (ng_icmpv6_hdr_t *)pkt->data;
+    dis = (ng_rpl_dis_t *)(icmp + 1);
+    dis->flags = 0;
+    dis->reserved = 0;;
+    memset((dis + 1), 0, 4);
+
```

Comment:
The `DIS` in this minimal form is too small when testing with native .. wireshark complains that the ethernet frame check seq. is incorrect. As a quick fix I added 4 `PAD1` options, which would currently also appear for other link layers. Should we tackle this with a follow-up PR and a little bit of discussion beforehand?